Vagrantfile

```diff
@@ -55,6 +55,17 @@ Vagrant.configure(2) do |config|
       domain.nested = true  # enable nested virtualization
       domain.cpu_mode = "host-model"
 
+      # We need 3 bricks, let us create 500G disks for that.
+      # NOTE: if size is changed, edit storage/gluster-deploy.sh
+      "bcd".split("").each do |i|
```

Comment:
Could you explain why, we need dedicated disks and can't have something completely inside the vm?

Reply:
Can we have a container, which is part of the manifest, which creates three qcow2 disks on deployment and mounts them as nbd devices instead?

Reply:
I don't understand where that would run, nor how that's any simpler/easier/cleaner than doing it in the Vagrantfile.

Reply:
The more we do inside the manifests, the easier it is to deploy it on differnt clusters. See for instance: https://github.com/rmohr/ceph-cinder-demo/blob/master/storage/templates/provisioner.yaml. This provisions ceph and cinder in a  single-node setup, without the need to touch the node in any way. For as long as a user can run provileged pods (for host network), this can be used in any k8s/openshift cluster.

That has a few advantages:
 * One does not need to rely on the "kubevirt" repo, just to develop with storage
 * One does not have create/change anything on the node to add storage, one just adds or removes manifests (no vagrant destroy, ...).
 * Of course that all is not for production, but we only need it to test code on top of gluster/cinder/ceph
...

I hope to see the same for gluster, so that we can simply add/remove gluster on arbitrary running clusters if needed.

Reply:
How do you creat qcoe2 images and attach them to the VM from inside the VM?

Reply:
Out of my head I know of two ways on how to create and attach the files from within a container:
 * loopback
 * nbd

Loopback works in principle like this inside a container:

```bash
truncate disk.img --size 2G
mknod /dev/loop9 -m0660 b 7 9
losetup /dev/loop9 disk.img
losetup -a
```

For nbd make sure that you mount "/lib/modules" to the container, then do something like this:

```bash
modprobe nbd max_part=8
mknod /dev/nbd0 b 43 0
qemu-nbd --connect=/dev/nbd0 file.qcow2
```

Reply:
Hmm... so this is done within the container and it changes the state of the host it's running on? And what are the differences, advantages, and/or disadvantages between the two methods?

Reply:
Maybe the loopback is more predictable and easier, since we don't need to load kernel modules.

It is a little bit hard for me to see what gk-deploy mechanism does with the topology, but I guess a gluster component is started on the node, which exposes these devices via a service? With that approach we can mount the devices inside the container (that this is also visible on the host is just a side-effect).

Is there then anything missing, to just provide a pre-created yaml manifest for development/testing without gk-deploy? Is there still something running on the node, or needs to be run on the node?

Reply:
gk-deploy does a fair number of things that can't be encapsulated in a single YAML manifest. Many of the resources deployed have to be done in a specific order and not all at once. You also can't pre-populate the heketi database, so the topology load, volume creation, and database copy have to be performed every time.

I would recommend setting up a meeting to get a run-down of what gk-deploy does, how the different components operate, and what requirements are mandatory vs. configuration-specific.

Reply:
> I would recommend setting up a meeting to get a run-down of what gk-deploy does, how the different components operate, and what requirements are mandatory vs. configuration-specific.

We can talk about it, sure.

> gk-deploy does a fair number of things that can't be encapsulated in a single YAML manifest. Many of the resources deployed have to be done in a specific order and not all at once.

That's the same for ceph and cinder. There are two very simple ways to make sure that the order is right:
1. Let containers exit with a non-zero exit code if it can't connect to a prerequisite service
2. Let containers wait for "ready" files to appear in a shared directory (emptyDir).

Reply:
Hmm... I'm not very well versed in container-thinking yet, it seems. I'd like to learn more about this. Probably in the same meeting as above? :)